db/wide/wide_column_serialization.cc

```diff
@@ -0,0 +1,162 @@
+//  (found in the LICENSE.Apache file in the root directory).
+
+#include "db/wide/wide_column_serialization.h"
+
+#include <algorithm>
+#include <cassert>
+
+#include "port/port.h"
+#include "util/autovector.h"
+#include "util/coding.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+Status WideColumnSerialization::Serialize(const WideColumnDescs& column_descs,
+                                          std::string* output) {
+  // Column names should be strictly ascending
+  assert(std::adjacent_find(
+             column_descs.cbegin(), column_descs.cend(),
+             [](const WideColumnDesc& lhs, const WideColumnDesc& rhs) {
+               return lhs.name().compare(rhs.name()) > 0;
+             }) == column_descs.cend());
+  assert(output);
+
+  if (column_descs.size() > static_cast<size_t>(port::kMaxUint32)) {
+    return Status::InvalidArgument("Too many wide columns");
+  }
+
+  PutVarint32(output, kCurrentVersion);
+
+  PutVarint32(output, static_cast<uint32_t>(column_descs.size()));
+
+  size_t total_column_value_size = 0;
+
+  for (const auto& desc : column_descs) {
+    const Slice& name = desc.name();
+    if (name.size() > static_cast<size_t>(port::kMaxUint32)) {
+      return Status::InvalidArgument("Wide column name too long");
+    }
+
+    const Slice& value = desc.value();
+    if (value.size() > static_cast<size_t>(port::kMaxUint32)) {
+      return Status::InvalidArgument("Wide column value too long");
+    }
+
+    PutLengthPrefixedSlice(output, name);
+    PutVarint32(output, static_cast<uint32_t>(value.size()));
+
+    total_column_value_size += value.size();
+  }
+
+  output->reserve(output->size() + total_column_value_size);
+
+  for (const auto& desc : column_descs) {
+    const Slice& value = desc.value();
+
+    output->append(value.data(), value.size());
+  }
+
+  return Status::OK();
+}
+
+Status WideColumnSerialization::DeserializeOne(Slice* input,
+                                               const Slice& column_name,
+                                               WideColumnDesc* column_desc) {
+  WideColumnDescs all_column_descs;
+
+  const Status s = DeserializeIndex(input, &all_column_descs);
+  if (!s.ok()) {
+    return s;
+  }
+
+  auto it = std::lower_bound(all_column_descs.cbegin(), all_column_descs.cend(),
+                             column_name,
+                             [](const WideColumnDesc& lhs, const Slice& rhs) {
+                               return lhs.name().compare(rhs) < 0;
+                             });
+  if (it == all_column_descs.end() || it->name() != column_name) {
+    return Status::NotFound("Wide column not found");
+  }
+
+  *column_desc = *it;
+
+  return Status::OK();
+}
+
+Status WideColumnSerialization::DeserializeAll(Slice* input,
+                                               WideColumnDescs* column_descs) {
+  return DeserializeIndex(input, column_descs);
+}
+
+Status WideColumnSerialization::DeserializeIndex(
```

Comment:
I'm confused why it is called DeserializeIndex(). It appears to deserialize everything.